default/python/AI/ResourcesAI.py

```diff
@@ -148,115 +187,338 @@ def check_avoid_change_due_to_bureaucracy(self, pinfo: PlanetFocusInfo, focus: s
         last_turn = fo.currentTurn() - 1
         if pinfo.planet.LastTurnColonized == last_turn or pinfo.planet.LastTurnConquered == last_turn:
             return False  # no penalty for newly settled/conquered planets
-        if pinfo.current_output[2] + 0.5 < pinfo.possible_output[pinfo.current_focus][2]:
+        if pinfo.current_output.influence + 0.5 < pinfo.possible_output[pinfo.current_focus].influence:
             # avoid repeated focus changes: When influence production already is below target, do not
             # get it further down by switching again (I've seen one planet at -30...)
             debug(f"Avoid focus change on {pinfo.planet.name} due bureaucracy and influence not maxed.")
             return True
-        threshold = 5.0
-        if self.production_value(pinfo, pinfo.current_focus) + threshold > self.production_value(pinfo, focus):
-            debug(f"Avoid focus change on {pinfo.planet.name} due bureaucracy and gain below {threshold}.")
-            return True
         return False
 
-    def production_value(self, pinfo: PlanetFocusInfo, focus: str) -> float:
+    def production_rating(self, pp: float, rp: float, ip: float, stability: float, focus: str) -> float:
```

Comment:
The name of the function is production but you mention protection twice.

Reply:
protection is one of the possible foci, whose output can is rated. Anyway, rate_output sound like a better name.